sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java

```diff
@@ -146,51 +152,93 @@ static Object loadYaml(InputStream inputStream, Map<String, String> environmentV
+     * performs environment variable substitution.
+     */
     @Override
+    @SuppressWarnings({"ReturnValueIgnored", "CatchingUnchecked"})
     protected Map<Object, Object> constructMapping(MappingNode node) {
-      // First call the super to construct mapping from MappingNode as usual
-      Map<Object, Object> result = super.constructMapping(node);
-
-      // Iterate through the map entries, and:
-      // 1. Identify entries which are scalar strings eligible for environment variable substitution
-      // 2. Apply environment variable substitution
-      // 3. Re-parse substituted value so it has correct type (i.e. yaml.load(newVal))
-      for (Map.Entry<Object, Object> entry : result.entrySet()) {
-        Object value = entry.getValue();
-        if (!(value instanceof String)) {
-          continue;
+      Map<Object, Object> mapping = settings.getDefaultMap().apply(node.getValue().size());
+      List<NodeTuple> nodeValue = node.getValue();
+      for (NodeTuple tuple : nodeValue) {
+        Node keyNode = tuple.getKeyNode();
+        Node valueNode = tuple.getValueNode();
+        Object key = constructObject(keyNode);
+        if (key != null) {
+          try {
+            key.hashCode(); // check circular dependencies
+          } catch (Exception e) {
+            throw new ConstructorException(
+                "while constructing a mapping",
+                node.getStartMark(),
+                "found unacceptable key " + key,
+                tuple.getKeyNode().getStartMark(),
+                e);
+          }
         }
-
-        String val = (String) value;
-        Matcher matcher = ENV_VARIABLE_REFERENCE.matcher(val);
-        if (!matcher.find()) {
-          continue;
+        Object value = constructValueObject(valueNode);
+        if (keyNode.isRecursive()) {
+          if (settings.getAllowRecursiveKeys()) {
+            postponeMapFilling(mapping, key, value);
+          } else {
+            throw new YamlEngineException(
+                "Recursive key for mapping is detected but it is not configured to be allowed.");
+          }
+        } else {
+          mapping.put(key, value);
         }
+      }
 
-        int offset = 0;
-        StringBuilder newVal = new StringBuilder();
-        do {
-          MatchResult matchResult = matcher.toMatchResult();
-          String replacement = environmentVariables.getOrDefault(matcher.group(1), "");
-          newVal.append(val, offset, matchResult.start()).append(replacement);
-          offset = matchResult.end();
-        } while (matcher.find());
-        if (offset != val.length()) {
-          newVal.append(val, offset, val.length());
-        }
-        entry.setValue(yaml.load(newVal.toString()));
+      return mapping;
+    }
+
+    private Object constructValueObject(Node node) {
+      if (!(node instanceof ScalarNode)) {
+        return super.constructObject(node);
```

Comment:
Don't really need that `super` here.

```suggestion
        return constructObject(node);
```

Reply:
Good catch. At one point the implementation overrode `constructObject` so super was necessary but I adjusted it in a revision.